### PR TITLE
Update Run Tasks for global config and stages

### PR DESCRIPTION
**Pull request by glennsarti**
This PR needs the Global Run Tasks feature to be generally available, and deployed to the CI environment before CI will pass

---

## Description

This commit also extracts the subscription update process and updated it as it
fails in the local development. Specifically, there may be 2 or more tokens
used to administer a TFC instance which have different admin scopes. This commit
now encapsulates that logic inside a SubscriptionUpdater "service" class which
can be used to build up the request to update a subscription.

This commit updates the entitlement set object to include the new
global-run-tasks feature.

The global run tasks feature updates both the Organization and Workspace
Run tasks with new attributes. Unfortunately these new attributes exposed
an issue with the JSON API pacakge, that it doesn't support complex objects
(Custom maps or non-standard types) [1]. Instead we need to unmarshall to
a more simpler object and then we adapt the simpler object to our more
correct struct.

[1] https://github.com/google/jsonapi/issues/74

## Testing plan

Updated integration tests exercise this behaviour.

## External links

- [API documentation](https://github.com/hashicorp/terraform-docs-common/pull/561)
- [JSON API issue](https://github.com/google/jsonapi/issues/74)

## Output from tests
Including output from tests may require access to a TFE instance. Ignore this section if you have no environment to test against.

```text
go test -v ./... --tags=integration --count=1 --run TestWorkspaceRunTasks
...
=== RUN   TestWorkspaceRunTasksCreate
=== RUN   TestWorkspaceRunTasksCreate/attach_run_task_to_workspace
=== RUN   TestWorkspaceRunTasksCreate/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly
--- PASS: TestWorkspaceRunTasksCreate (6.15s)
    --- PASS: TestWorkspaceRunTasksCreate/attach_run_task_to_workspace (0.81s)
        --- PASS: TestWorkspaceRunTasksCreate/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly (0.00s)
=== RUN   TestWorkspaceRunTasksCreateDeprecated
=== RUN   TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace
=== RUN   TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly
--- PASS: TestWorkspaceRunTasksCreateDeprecated (5.63s)
    --- PASS: TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace (0.78s)
        --- PASS: TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly (0.00s)
=== RUN   TestWorkspaceRunTasksList
=== RUN   TestWorkspaceRunTasksList/with_no_params
--- PASS: TestWorkspaceRunTasksList (7.93s)
    --- PASS: TestWorkspaceRunTasksList/with_no_params (0.40s)
=== RUN   TestWorkspaceRunTasksRead
=== RUN   TestWorkspaceRunTasksRead/by_ID
=== RUN   TestWorkspaceRunTasksRead/by_ID/ensure_run_task_is_deserialized
=== RUN   TestWorkspaceRunTasksRead/by_ID/ensure_workspace_is_deserialized
--- PASS: TestWorkspaceRunTasksRead (5.93s)
    --- PASS: TestWorkspaceRunTasksRead/by_ID (0.43s)
        --- PASS: TestWorkspaceRunTasksRead/by_ID/ensure_run_task_is_deserialized (0.00s)
        --- PASS: TestWorkspaceRunTasksRead/by_ID/ensure_workspace_is_deserialized (0.00s)
=== RUN   TestWorkspaceRunTasksUpdate
=== RUN   TestWorkspaceRunTasksUpdate/update_task
--- PASS: TestWorkspaceRunTasksUpdate (6.43s)
    --- PASS: TestWorkspaceRunTasksUpdate/update_task (0.73s)
=== RUN   TestWorkspaceRunTasksUpdateDeprecated
=== RUN   TestWorkspaceRunTasksUpdateDeprecated/update_task
--- PASS: TestWorkspaceRunTasksUpdateDeprecated (6.24s)
    --- PASS: TestWorkspaceRunTasksUpdateDeprecated/update_task (0.74s)
=== RUN   TestWorkspaceRunTasksDelete
=== RUN   TestWorkspaceRunTasksDelete/with_valid_options
=== RUN   TestWorkspaceRunTasksDelete/when_the_workspace_run_task_does_not_exist
=== RUN   TestWorkspaceRunTasksDelete/when_the_workspace_does_not_exist
--- PASS: TestWorkspaceRunTasksDelete (6.63s)
    --- PASS: TestWorkspaceRunTasksDelete/with_valid_options (0.71s)
    --- PASS: TestWorkspaceRunTasksDelete/when_the_workspace_run_task_does_not_exist (0.25s)
    --- PASS: TestWorkspaceRunTasksDelete/when_the_workspace_does_not_exist (0.27s)
```

```text
go test -v ./... --tags=integration --count=1 --run TestRunTasks
...
=== RUN   TestWorkspaceRunTasksCreate
=== RUN   TestWorkspaceRunTasksCreate/attach_run_task_to_workspace
=== RUN   TestWorkspaceRunTasksCreate/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly
--- PASS: TestWorkspaceRunTasksCreate (5.94s)
    --- PASS: TestWorkspaceRunTasksCreate/attach_run_task_to_workspace (0.79s)
        --- PASS: TestWorkspaceRunTasksCreate/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly (0.00s)
=== RUN   TestWorkspaceRunTasksCreateDeprecated
=== RUN   TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace
=== RUN   TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly
--- PASS: TestWorkspaceRunTasksCreateDeprecated (5.73s)
    --- PASS: TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace (0.86s)
        --- PASS: TestWorkspaceRunTasksCreateDeprecated/attach_run_task_to_workspace/ensure_run_task_is_deserialized_properly (0.00s)
=== RUN   TestWorkspaceRunTasksList
=== RUN   TestWorkspaceRunTasksList/with_no_params
--- PASS: TestWorkspaceRunTasksList (7.66s)
    --- PASS: TestWorkspaceRunTasksList/with_no_params (0.41s)
=== RUN   TestWorkspaceRunTasksRead
=== RUN   TestWorkspaceRunTasksRead/by_ID
=== RUN   TestWorkspaceRunTasksRead/by_ID/ensure_run_task_is_deserialized
=== RUN   TestWorkspaceRunTasksRead/by_ID/ensure_workspace_is_deserialized
--- PASS: TestWorkspaceRunTasksRead (5.96s)
    --- PASS: TestWorkspaceRunTasksRead/by_ID (0.39s)
        --- PASS: TestWorkspaceRunTasksRead/by_ID/ensure_run_task_is_deserialized (0.00s)
        --- PASS: TestWorkspaceRunTasksRead/by_ID/ensure_workspace_is_deserialized (0.00s)
=== RUN   TestWorkspaceRunTasksUpdate
=== RUN   TestWorkspaceRunTasksUpdate/update_task
--- PASS: TestWorkspaceRunTasksUpdate (6.60s)
    --- PASS: TestWorkspaceRunTasksUpdate/update_task (0.80s)
=== RUN   TestWorkspaceRunTasksUpdateDeprecated
=== RUN   TestWorkspaceRunTasksUpdateDeprecated/update_task
--- PASS: TestWorkspaceRunTasksUpdateDeprecated (6.24s)
    --- PASS: TestWorkspaceRunTasksUpdateDeprecated/update_task (0.74s)
=== RUN   TestWorkspaceRunTasksDelete
=== RUN   TestWorkspaceRunTasksDelete/with_valid_options
=== RUN   TestWorkspaceRunTasksDelete/when_the_workspace_run_task_does_not_exist
=== RUN   TestWorkspaceRunTasksDelete/when_the_workspace_does_not_exist
--- PASS: TestWorkspaceRunTasksDelete (6.46s)
    --- PASS: TestWorkspaceRunTasksDelete/with_valid_options (0.69s)
    --- PASS: TestWorkspaceRunTasksDelete/when_the_workspace_run_task_does_not_exist (0.24s)
    --- PASS: TestWorkspaceRunTasksDelete/when_the_workspace_does_not_exist (0.24s)
PASS
```

### CODE REVIEW

**Comment by glennsarti** [comment]
Add Compliance team for review. Not that it's required, but a nice to have

File: run_task.go [hunk 1]
```diff
@@ -87,6 +95,13 @@ type RunTaskReadOptions struct {
 	Include []RunTaskIncludeOpt `url:"include,omitempty"`
 }
 
+// GlobalRunTask represents the optional global configuration of a TFC/E run task
```
Comment (mrinalirao) [inline run_task.go:98]:
```suggestion
// GlobalRunTaskOptions represents the optional global configuration of a TFC/E run task
```

File: helper_test.go [hunk 2]
```diff
@@ -2526,49 +2500,9 @@ func createVariableSetVariable(t *testing.T, client *Client, vs *VariableSet, op
...
-		return
-	}
-
-	err = req.Do(context.Background(), nil)
-	if err != nil {
-		t.Fatalf("Failed to upgrade subscription: %v", err)
-	}
+// DEPRECATED : Please use the newSubscriptionUpdater instead.
+func upgradeOrganizationSubscription(t *testing.T, _ *Client, organization *Organization) {
+	newSubscriptionUpdater(organization).WithBusinessPlan().Update(t)
```
Comment (mrinalirao) [inline helper_test.go:2505]: nice!! 👍

**Review by mrinalirao — DISMISSED** [review]
A minor comment fix. LGTM otherwise! 👍

File: run_task.go (hunk 1 above)
Comment (glennsarti) [reply run_task.go]: @mrinalirao Will need a +1 again.  Sorry

File: internal_run_task.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+// A private struct we need for unmarshalling
+type internalRunTask struct {
+	ID          string                 `jsonapi:"primary,tasks"`
+	Name        string                 `jsonapi:"attr,name"`
+	URL         string                 `jsonapi:"attr,url"`
+	Description string                 `jsonapi:"attr,description"`
+	Category    string                 `jsonapi:"attr,category"`
+	HMACKey     *string                `jsonapi:"attr,hmac-key,omitempty"`
+	Enabled     bool                   `jsonapi:"attr,enabled"`
+	RawGlobal   map[string]interface{} `jsonapi:"attr,global-configuration,omitempty"`
```
Comment (brandonc) [inline internal_run_task.go:15]: TIL that jsonapi doesn't automatically decode json annotated attribute object types. It's certainly in spec to have an object within an attribute. I will take a closer look after I finish this review.

**Review by brandonc — APPROVED** [review]
Very nicely done to expose the interface you want and hide the problematic object case. I also appreciate the subscription update cleanup

File: internal_run_task.go (hunk 3 above)
Comment (glennsarti) [reply internal_run_task.go:15]: Unfortunately, jsonapi is effectively a dead project :-( It had a "No Maintenance Intended" badge added 4 years ago and the PR/Issues list is wasteland. We'd have to fork it if we wanted to fix any issues with it.

**Pull request merged**